File: capability/basics/sessions.py
"""会话管理功能 — 列出、匹配、恢复、加载会话。

所有函数接受 MemoryProvider 作为参数，返回结构化数据。
"""

from __future__ import annotations

from memory import MemoryProvider


def list_sessions(memory: MemoryProvider) -> list[dict]:
    """返回所有 session 元数据（按更新时间倒序）。

    每个 dict 包含: id, created, updated, message_count, preview
    """
    return memory.list_sessions()


def get_session_by_prefix(memory: MemoryProvider, prefix: str) -> dict | None | str:
    """按 ID 前缀匹配 session。

    返回:
        dict    — 唯一匹配的 session 元数据
        None    — 无匹配
        "ambiguous" — 多个匹配
    """
    sessions = memory.list_sessions()
    matches = [s for s in sessions if s["id"].startswith(prefix)]
    if len(matches) == 1:
        return matches[0]
    if len(matches) > 1:
        return "ambiguous"
    return None


def resume_session(memory: MemoryProvider, session_id: str) -> list[dict]:
    """恢复 session 并返回完整消息列表。

    每个 dict 包含: role, content, timestamp
    FileNotFoundError — session 文件不存在
    """
    return memory.resume_session(session_id)


def load_session_messages(memory: MemoryProvider, session_id: str) -> list[dict]:
    """加载 session 消息（只读，不切换当前 session）。"""
    return memory.load_session(session_id)

File: src/capability/basics/sessions.ts
/**
 * 会话管理工具 — 列出、匹配、恢复会话
 */

import type { MemoryProvider } from '../../memory/types.js'
import type { SessionIndexEntry } from '../../types.js'

/**
 * 列出所有会话索引
 */
export function listSessions(memory: MemoryProvider): SessionIndexEntry[] {
  return memory.listSessions()
}

/**
 * 按前缀匹配唯一会话；多个匹配或无匹配时返回 null
 */
export function get_session_by_prefix(memory: MemoryProvider, prefix: string): SessionIndexEntry | null {
  const sessions = memory.listSessions()
  const matches = sessions.filter(s => s.id.startsWith(prefix))
  return matches.length === 1 ? matches[0] : null
}

/**
 * 恢复指定会话；sessionId 无效时返回 false
 */
export async function resumeSession(memory: MemoryProvider, sessionId: string): Promise<boolean> {
  const sessions = memory.listSessions()
  const session = sessions.find(s => s.id === sessionId)
  if (!session) return false
  await memory.resumeSession(sessionId)
  return true
}
